kernel/pipes.c

```diff
@@ -457,6 +484,16 @@ int z_impl_k_pipe_put(struct k_pipe *pipe, void *data, size_t bytes_to_write,
 
 	z_sched_wait(&pipe->lock, key, &pipe->wait_q.writers, timeout, NULL);
 
+	/*
+	 * On SMP systems, threads in the processing list may timeout before
+	 * the data has finished copying. The following spin lock/unlock pair
+	 * prevents those threads from executing further until the data copying
+	 * is complete.
+	 */
+
+	key = k_spin_lock(&pipe->lock);
```

Comment:
Seems like this is being used to spin and wait for a completion flag effectively, spin locks have some implied irq mask/unmask operations tied to them, do you need that behavior as well or only an atomic cas loop? The comment above doesn't help in terms of the arch irq locking behavior being desired/undesired.

Reply:
I'm not the happiest with using the spin lock here, but it seemed to be the simplest mechanism to accomplish the following two goals.

 1.  Serve as an indicator that it is safe to continue execution.
 2. Prevent the thread from executing while another core was still in the process of copying the data that the thread had requested.

The biggest downside to this approach that I can see is (and admittedly you may see something more) that the one core is essentially stalled while the other core finishes copying data. However, as an attempt to put this into perspective, I pose the following question ... "Is blocking on the spin lock on the 'tail end' of the pipe operation fundamentally any worse than blocking on a spin lock at the 'front end' of the pipe operation?"